### PR TITLE
Port to Python 3

### PR DESCRIPTION
**Pull request by hawkowl**
Since this needs to run on Python 3, I've removed the Python 3 linter checks -- since it won't be able to run them, and the Twisted codebase is now free of the majority of them.

This also updates to the Py3 compatible version of pylint and newer PEP8.

### CODE REVIEW

**Comment by hawkowl** [comment]
cc @glyph :)

File: twistedchecker/test/test_runner.py [hunk 1]
```diff
@@ -77,14 +82,13 @@ def test_findUselessCheckers(self):
         Test for method findUselessCheckers
         """
         runner = Runner()
-        registeredCheckers = sum(runner.linter._checkers.values(), [])
+        registeredCheckers = sum(list(runner.linter._checkers.values()), [])
```
Comment (rodrigc) [inline twistedchecker/test/test_runner.py]: Do you really need to explicitly call list() in these changes,
or can you leave the calls to list() out?  converters like 2to3 tend to put them in, but most times you don't need them.

**Comment by glyph** [comment]
Use of the word "also" is a killer :).  Can you split this up?  Particularly, the verbose pep8=>pycodestyle renaming detracts from trying to look at the rest of the code.

**Comment by glyph** [comment]
One other minor concern - I'd really like to be able to run twistedchecker locally using pypy to get faster lint times.  Right now I can't, due to this unfortunate bug - https://github.com/PyCQA/pylint/issues/964 - but if we stop testing twistedchecker on 2.7 then this is even less likely to be possible.

**Comment by glyph** [comment]
D'oh. I suppose you can remove that last comment.  I thought I had seen somewhere that this was removing the 2.7 support - but it looks like the tox envs just add 3.5.

File: setup.py [hunk 2]
```diff
@@ -9,7 +9,7 @@
 setup(
     name='TwistedChecker',
     description='A Twisted coding standard compliance checker.',
-    version='0.5.0',
+    version='16.0',
```
Comment (glyph) [inline setup.py]: This seems pretty clearly unrelated…

File: setup.py [hunk 3]
```diff
@@ -30,23 +30,25 @@
         "License :: OSI Approved :: MIT License",
         "Programming Language :: Python",
         "Programming Language :: Python :: 2",
-        "Programming Language :: Python :: 2.6",
```
Comment (glyph) [inline setup.py]: yaaay

File: setup.py [hunk 4]
```diff
@@ -30,23 +30,25 @@
...
     extras_require = {
         'dev':  [
             'twisted>=15.0.0',
-            'pyflakes==0.8.1',
+            'pyflakes',
+            'coverage'
             ],
     },
-    long_description=file('README.rst').read()
+    long_description=open('README.rst').read()
```
Comment (glyph) [inline setup.py]: `ResourceWarning` :)

File: tox.ini [hunk 5]
```diff
@@ -24,6 +24,7 @@ deps =
 
 commands =
    python -V
+   pip install '.[dev]'
```
Comment (glyph) [inline tox.ini]: Why is this necessary now where it wasn't before?

File: setup.py [hunk 6]
```diff
@@ -30,23 +30,25 @@
...
-        "Programming Language :: Python :: 2.6",
         "Programming Language :: Python :: 2.7",
+        "Programming Language :: Python :: 3",
+        "Programming Language :: Python :: 3.5",
         "Topic :: Software Development :: Quality Assurance"
     ],
     keywords=[
         "twisted", "checker", "compliance", "pep8"
     ],
     install_requires=[
```
Comment (glyph) [inline setup.py]: These version bumps should be in an earlier PR, probably.

File: twistedchecker/checkers/comment.py [hunk 7]
```diff
@@ -13,11 +13,13 @@ class CommentChecker(BaseChecker):
     """
     msgs = {
      'W9401': ('Comments should begin with one whitespace',
-               'Used for checking comment format issues.'),
+               'Used for checking comment format issues.',
+                   'comments-one-whitespace'),
```
Comment (glyph) [inline twistedchecker/checkers/comment.py]: Is this change in format related to the upgrade of pylint?

File: twistedchecker/checkers/comment.py [hunk 8]
```diff
@@ -13,11 +13,13 @@ class CommentChecker(BaseChecker):
...
      'W9402': ('The first letter of comment should be capitalized',
-               'Used for checking comment format issues.')
+               'Used for checking comment format issues. ',
+                   'comments-capitalized')
     }
-    __implements__ = IASTNGChecker
+    __implements__ = IAstroidChecker
```
Comment (glyph) [inline twistedchecker/checkers/comment.py]: someone should tell pylint the good news about ZI…

File: twistedchecker/checkers/comment.py [hunk 9]
```diff
@@ -32,9 +34,9 @@ def visit_module(self, node):
             return
         isFirstLineOfComment = True
         isDocString = False
-        lines = node.file_stream.readlines()
+        lines = node.stream().readlines()
```
Comment (glyph) [inline twistedchecker/checkers/comment.py]: is this also a spot for a `ResourceWarning`?

File: twistedchecker/checkers/pycodestyleformat.py [hunk 10]
```diff
@@ -158,16 +160,16 @@ class PEP8Checker(BaseChecker):
...
     def __init__(self, linter):
         """
-        Change function of processing blank lines in pep8.
+        Change function of processing blank lines in pycodestyle.
 
         @param linter: current C{PyLinter} object.
         """
         BaseChecker.__init__(self, linter)
-        self.pep8Enabled = self.linter.option_value("pep8")
+        self.pycodestyleEnabled = True # yes of course
```
Comment (glyph) [inline twistedchecker/checkers/pycodestyleformat.py]: why did this change?

File: twistedchecker/checkers/pycodestyleformat.py [hunk 11]
```diff
@@ -176,22 +178,13 @@ def visit_module(self, node):
...
-        @param file: path of module to check
-        """
-        recorder = PEP8WarningRecorder(file)
-        self._outputMessages(recorder.warnings)
+        recorder = PyCodeStyleWarningRecorder(node.file)
+        self._outputMessages(recorder.warnings, node)
 
 
-    def _outputMessages(self, warnings):
+    def _outputMessages(self, warnings, node):
```
Comment (glyph) [inline twistedchecker/checkers/pycodestyleformat.py]: `@type node`

File: twistedchecker/checkers/python3.py [hunk 12]
```diff
@@ -1,187 +0,0 @@
-"""
-Checker for python3 compatibility issues.
```
Comment (glyph) [inline twistedchecker/checkers/python3.py]: There are projects other than Twisted which use twistedchecker (axiom and vertex come to mind) which are not finished porting to python 3, so this seems premature.

File: twistedchecker/configuration/pylintrc [hunk 13]
```diff
@@ -75,9 +75,7 @@ variable-rgx=([a-z_][a-zA-Z0-9]*)$
 # class names, so just make sure they are valid.
 const-rgx=(([A-Z_][A-Z0-9_]*)|(__.*__)|([a-z_][a-zA-Z0-9]+)|([A-Z][a-zA-Z0-9]+))$
 
-# To prevent generating two warnings when found an invalid
-# argument name, disable argument name checking.
-argument-rgx=.*$
+argument-rgx=([a-z_][a-zA-Z0-9]*)$
```
Comment (glyph) [inline twistedchecker/configuration/pylintrc]: What does re-enabling this have to do with porting to python 3?

File: twistedchecker/core/runner.py [hunk 14]
```diff
@@ -145,7 +145,7 @@ def registerCheckers(self):
                                         fromlist=["twistedchecker.checkers"]),
                              classname)
             instanceChecker = checker(self.linter)
-            allowedMessages += instanceChecker.msgs.keys()
+            allowedMessages += list(instanceChecker.msgs.keys())
```
Comment (glyph) [inline twistedchecker/core/runner.py]: Perhaps this should be changed to `.extend` instead of `list`?

File: twistedchecker/functionaltests/maxlinelength.py [hunk 15]
```diff
@@ -1,6 +1,6 @@
 # enable: C0301
 # max line length of Twsited projects should be 79
-print "this line is long long  long  long long long long long ends at column 80"
+print("this line is long long  long  long long long long long ends at column 80")
```
Comment (glyph) [inline twistedchecker/functionaltests/maxlinelength.py]: It doesn't though - it ends at column 81. You removed one space and added two parentheses.  The fact that this doesn't cause anything to fail is cause for alarm :)

File: twistedchecker/test/test_limitedreporter.py [hunk 16]
```diff
@@ -1,5 +1,9 @@
 import os
-import StringIO
+
+try:
+    from StringIO import StringIO
+except ImportError:
+    from io import StringIO
```
Comment (glyph) [inline twistedchecker/test/test_limitedreporter.py]: Why does this need to be conditional?  Can't it be `io.StringIO` on both versions?

**Comment by glyph** [comment]
Reviewed!

@hawkowl - Thank you for taking on the port here.  But the change is way too big and does a bunch of unrelated things.  Please split this into a series of smaller changes so that each can be reviewed more quickly and landed independently.  Particularly, I'd like to see the version bumps taken care of first without the rest of the py3 port involved.

File: setup.py [hunk 17]
```diff
@@ -30,23 +30,25 @@
...
-        "pylint == 0.26.0",
-        "logilab-common == 0.62.0",
-        "pep8 == 1.5.7"
+        "pylint == 1.5.6",
+        "logilab-common == 1.2.1",
```
Comment (PCManticore) [inline setup.py]: logilab-common is not needed anymore by pylint ;-)

File: twistedchecker/checkers/python3.py (hunk 12 above)
Comment (PCManticore) [inline twistedchecker/checkers/python3.py]: By the way, there is also pylint --py3k which you might use for the same problem. It seems more feature complete than what you are currently using, maybe it is worth taking a look.

Comment (PCManticore) [inline twistedchecker/checkers/python3.py]: Although we don't have the ``except Exception, e`` check. Would be really cool if this would be included upstream. ;-)

File: setup.py (hunk 17 above)
Comment (hawkowl) [inline setup.py]: Except we use it somewhere, yaaaaaay :P

File: setup.py (hunk 6 above)
Comment (hawkowl) [inline setup.py]: The problem is that bumping the version explodes everything, because pylint has no such thing as API stability ;)

This PR is actually "Upgrade the PyLint to a version that actually works on Python 3", with some "make things work on Python 3" as well :)

File: setup.py (hunk 17 above)
Comment (glyph) [inline setup.py]: This is extra-confusing then. Aren't we using it to import the exception types and such required by pylint?  If pylint doesn't use them any more, can we just delete that code?

File: twistedchecker/checkers/comment.py (hunk 7 above)
Comment (hawkowl) [inline twistedchecker/checkers/comment.py]: Yep!

File: twistedchecker/checkers/python3.py (hunk 12 above)
Comment (hawkowl) [inline twistedchecker/checkers/python3.py]: @glyph Okay, I can revert this (and disable it if running Python 3, where the `ast` module explodes if you try and import code that violates some of these rules).

File: twistedchecker/configuration/pylintrc (hunk 13 above)
Comment (hawkowl) [inline twistedchecker/configuration/pylintrc]: This is a pylint upgrade thing :(

File: twistedchecker/checkers/python3.py (hunk 12 above)
Comment (glyph) [inline twistedchecker/checkers/python3.py]: Sounds good. Anything to make this monster diff smaller :-\

File: twistedchecker/test/test_runner.py (hunk 1 above)
Comment (rodrigc) [inline twistedchecker/test/test_runner.py]: On second look, these explicit calls to list() all look needed.
